#166272341 Rate an Article

EMacco opened the pull request:
#### What does this PR do?
Allows users to be able to rate an article

#### Description of Task to be completed?
    - Write unit tests
    - Document feature using swagger
    - Authenticate user before rating
    - Check if article exists
    - Validate the rate (1, 2, 3, 4, 5)
    - Update user rating when multiple rate is submitted

#### How should this be manually tested?
    - pull this branch
    - run "npm install"
    - run "npm test"
or run npm start:dev and send POST request to `http://localhost:3000/api/v1/articles/rate` using Postman

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
#166272341

#### Screenshots (if appropriate)
<img width="1368" alt="Screenshot 2019-06-11 at 10 23 57 PM" src="https://user-images.githubusercontent.com/20377428/59315044-e58e8980-8caf-11e9-8eff-b99826727143.png">


#### Questions:
N/A

EMacco committed:
feat-[166272341]: Rate Article

- Write unit tests
- Document feature using swagger
- Authenticate user before rating
- Check if article exists
- Validate the rate (1, 2, 3, 4, 5)
- Update user rating when multiple rate is submitted

[Finishes #166272341]

EMacco committed:
feat-[166272341]: Rate Article

- Write unit tests
- Document feature using swagger
- Authenticate user before rating
- Check if article exists
- Validate the rate (1, 2, 3, 4, 5)
- Update user rating when multiple rate is submitted

[Finishes #166272341]

EMacco committed:
Merge branch '166272341-ft-rate-article' of https://github.com/andela/kingsmen-ah-backend into 166272341-ft-rate-article

EMacco committed:
Merge branch '166272341-ft-rate-article' of https://github.com/andela/kingsmen-ah-backend into 166272341-ft-rate-article

EMacco committed:
Merge branch '166272341-ft-rate-article' of https://github.com/andela/kingsmen-ah-backend into 166272341-ft-rate-article

EMacco committed:
feat-[166272341]: Rate Article

    - Write unit tests
    - Document feature using swagger
    - Authenticate user before rating
    - Check if article exists
    - Validate the rate (1, 2, 3, 4, 5)
    - Update user rating when multiple rate is submitted

    [Finishes #166272341]

EMacco committed:
Merge branch 'develop' into 166272341-ft-rate-article

adex001 commented on package.json:
```diff
@@ -37,6 +37,7 @@
     "debug": "^4.1.1",
     "dotenv": "^8.0.0",
     "express": "^4.17.1",
+    "faker": "^4.1.0",
```
Since this is not used in the main package, I suggest you include it in the devDependencies.

fidelisojeah commented on server/controllers/articles.js:
```diff
@@ -0,0 +1,103 @@
+          {
+            model: User,
+            as: 'author',
+            attributes: { exclude: ['password', 'updatedAt', 'createdAt'] }
```
please include only the fields you need here rather than excluding the ones you do not need

fidelisojeah commented on server/controllers/articles.js:
```diff
@@ -0,0 +1,103 @@
+      // Get the full details of user and article
+      const moreDetails = await Article.findOne({
+        where: { id: articleId },
+        attributes: { exclude: ['updatedAt', 'createdAt'] },
```
you don't need to exclude this per se

fidelisojeah commented on server/controllers/articles.js:
```diff
@@ -0,0 +1,103 @@
+      if (!article) return Response.error(res, 404, 'Article does not exist');
+
+      // Get the full details of user and article
+      const moreDetails = await Article.findOne({
```
wait.

This is totally not needed.

You already have an article object on line 33

use that

adex001 commented on server/controllers/articles.js:
```diff
@@ -0,0 +1,103 @@
+      // Validate the rating
+      const articleDetails = await validateRating(req.body);
+
+      const { articleId, rate } = articleDetails;
```
Nice job! The article-id here seems to be provided by the user. Why not include it as a param to save the user from finding the article-id before rating.

fidelisojeah commented on server/controllers/articles.js:
```diff
@@ -0,0 +1,103 @@
+        // Create new rating for the user
+        const id = Math.floor(Math.random() * 1000000000) + 1;
+        newRating = await Rating.create({
+          id,
```
change the Rating model to make use of `uuid` for it's id field

fidelisojeah commented on server/controllers/articles.js:
```diff
@@ -0,0 +1,103 @@
+      let fullRatingInfo = { article: moreDetails };
+
+      // Check if this user has made a rating before
+      const userArticleRating = await Rating.findOne({
```
there may be other ways to do what you are trying to do here.

it may mean updating the `Rating` model however

fidelisojeah commented on server/helpers/Response.js:
```diff
@@ -18,6 +18,21 @@ class Response {
+  static success(res, code, data) {
+    return res.status(code).json({
+      status: code,
+      data
```
call this payload

fidelisojeah reviewed (changes requested):
I have updated the PT story.

reach out to me if you do not understand something

You may need to totally rework on this ticket.

EMacco committed:
Merge branch 'develop' into 166272341-ft-rate-article

EMacco committed:
feat-[166272341]: Rate Article

- Create route /articles/:articleId/rate
- Write unit tests
- Authenticate user before rating
- Check if article exists
- Validate the rate (1, 2, 3, 4, 5)
- Update user rating when multiple rate is submitted
- Document feature using swagger

[Finishes #166272341]

EMacco committed:
feat-[166272341]: Rate Article

- Create route /articles/:articleId/rate
- Write unit tests
- Authenticate user before rating
- Check if article exists
- Validate the rate (1, 2, 3, 4, 5)
- Update user rating when multiple rate is submitted
- Document feature using swagger

[Finishes #166272341]

fidelisojeah commented on server/controllers/articles.js:
```diff
@@ -27,67 +27,59 @@ class ArticleController {
-          userId,
-          articleId: req.body.articleId
-        }
+      const [newRating, created] = await Rating.findOrCreate({
```
what do you think is the point of having `associations` in sequelize if you gonna be doing stuff like this?

article.`set/add/get/etc`ratings (one or more of these methods is available to you) USE THEM

EMacco committed:
Merge branch '166272341-ft-rate-article' of https://github.com/andela/kingsmen-ah-backend into 166272341-ft-rate-article

EMacco committed:
Merge branch '166272341-ft-rate-article' of https://github.com/andela/kingsmen-ah-backend into 166272341-ft-rate-article

EMacco committed:
Merge branch '166272341-ft-rate-article' of https://github.com/andela/kingsmen-ah-backend into 166272341-ft-rate-article

EMacco committed:
Merge branch '166272341-ft-rate-article' of https://github.com/andela/kingsmen-ah-backend into 166272341-ft-rate-article

emmygozi reviewed (commented):
consider renaming these routes:

`server/tests/factory/article-factory.js ` and `server/tests/factory/user-factory.js ` to camel case in line with convention.

emmygozi commented on server/tests/factory/user-factory.js:
```diff
@@ -0,0 +1,23 @@
+import faker from 'faker';
+import Token from '../../helpers/Token';
```
consider adding @helper/ to replace '../../helpers/Token'

EMacco committed:
Merge branch '166272341-ft-rate-article' of https://github.com/andela/kingsmen-ah-backend into 166272341-ft-rate-article

EMacco committed:
feat-[166272341]: Rate Article

    - Create route /articles/:articleId/rate
    - Write unit tests
    - Authenticate user before rating
    - Check if article exists
    - Validate the rate (1, 2, 3, 4, 5)
    - Update user rating when multiple rate is submitted
    - Document feature using swagger

    [Finishes #166272341]

EMacco committed:
feat-[166272341]: Rate Article

    - Create route /articles/:articleId/rate
    - Write unit tests
    - Authenticate user before rating
    - Check if article exists
    - Validate the rate (1, 2, 3, 4, 5)
    - Update user rating when multiple rate is submitted
    - Document feature using swagger

    [Finishes #166272341]

EMacco committed:
Merge branch '166272341-ft-rate-article' of https://github.com/andela/kingsmen-ah-backend into 166272341-ft-rate-article

EMacco committed:
feat-[166272341]: Rate Article

        - Create route /articles/:articleId/rate
        - Write unit tests
        - Authenticate user before rating
        - Check if article exists
        - Validate the rate (1, 2, 3, 4, 5)
        - Update user rating when multiple rate is submitted
        - Document feature using swagger

        [Finishes #166272341]

fidelisojeah commented on server/tests/factory/userFactory.js:
```diff
@@ -0,0 +1,23 @@
+import faker from 'faker';
```
please the already created `user-factory`;

fidelisojeah commented on server/models/article.js:
```diff
@@ -34,7 +34,8 @@ module.exports = (sequelize, DataTypes) => {
+      foreignKey: 'userId',
+      as: 'author',
     });
 
```
```suggestion
Article.belongsToMany(User, {
      through: Rating,
      as: 'Ratings',
      foreignKey: 'articleId',
    });
```

fidelisojeah commented on server/models/rating.js:
```diff
@@ -1,5 +1,11 @@
 module.exports = (sequelize, DataTypes) => {
```
No associations should be defined in this file

fidelisojeah commented on server/controllers/articles.js:
```diff
@@ -0,0 +1,91 @@
+
+      // Check if the article exists
+      const article = await Article.findOne({
+        where: { id: articleId },
```
If I recall, on the readme, it uses the `slug` to identify the article

fidelisojeah commented on server/controllers/articles.js:
```diff
@@ -0,0 +1,91 @@
+      let result;
+
+      // Check if there is a rating
+      const userRating = await article.getRating({ where: { userId } });
```
Totally unneeded

```suggestion
    const result = await article.addRating(userId, { through: { ratings: rate }});
```

fidelisojeah commented on server/controllers/articles.js:
```diff
@@ -0,0 +1,91 @@
+      const newRating = { userId, ratings: rate };
+
+      let statusCode;
+      if (userRating) {
```
please remove all of these

EMacco replied:
I appreciate all the feedback. I will implement them immediately.
Thanks for your support

fidelisojeah reviewed (changes requested):
You should include the average ratings of an article as part of it's values

EMacco committed:
Merge branch 'develop' into 166272341-ft-rate-article

EMacco committed:
feat-[166272341]: Rate Article

    - Create route /articles/:articleId/rate
    - Write unit tests
    - Authenticate user before rating
    - Check if article exists
    - Validate the rate (1, 2, 3, 4, 5)
    - Update user rating when multiple rate is submitted
    - Document feature using swagger

    [Finishes #166272341]

EMacco committed:
feat-[166272341]: Rate Article

    - Create route /articles/:articleId/rate
    - Write unit tests
    - Authenticate user before rating
    - Check if article exists
    - Validate the rate (1, 2, 3, 4, 5)
    - Update user rating when multiple rate is submitted
    - Document feature using swagger

    [Finishes #166272341]

EMacco committed:
Merge branch '166272341-ft-rate-article' of https://github.com/andela/kingsmen-ah-backend into 166272341-ft-rate-article

adex001 commented on server/controllers/articles.js:
```diff
@@ -0,0 +1,109 @@
+      throw new Error('Parameters undefined');
+    }
+
+    return Article.findOne({
```
Article.findOne might return a promise here. try using the await with it.

EMacco replied:
Thanks for your feedback, but using await in this case will become redundant on a return value

adex001 commented on server/models/user.js:
```diff
@@ -73,9 +73,11 @@ module.exports = (sequelize, DataTypes) => {
     });
-    User.hasMany(Article, {
+
+    User.belongsToMany(Article, {
```
It's all cool to this point as I don't understand why a many-many association is used here. Can you explain further to me?

adex001 reviewed (dismissed):
use await in the findOne function

fidelisojeah commented on server/migrations/20190604161719-create-rating.js:
```diff
@@ -1,9 +1,10 @@
 module.exports = {
   up: (queryInterface, Sequelize) => queryInterface.createTable('Ratings', {
     id: {
```
remove `id` from ratings all together

EMacco replied:
Noted with thanks

fidelisojeah commented:
> consider renaming these routes:
> 
> `server/tests/factory/article-factory.js ` and `server/tests/factory/user-factory.js ` to camel case in line with convention.

don't do this please

EMacco committed:
feat-[166272341]: Rate Article

  - Create route /articles/:articleId/rate
  - Write unit tests
  - Authenticate user before rating
  - Check if article exists
  - Validate the rate (1, 2, 3, 4, 5)
  - Update user rating when multiple rate is submitted
  - Document feature using swagger
  - Format response object

  [Finishes #166272341]

EMacco committed:
Merge branch 'develop' into 166272341-ft-rate-article

EMacco committed:
Merge branch 'develop' into 166272341-ft-rate-article

EMacco committed:
Merge branch '166272341-ft-rate-article' of https://github.com/andela/kingsmen-ah-backend into 166272341-ft-rate-article

fidelisojeah reviewed (approved):
Good Job

emmygozi reviewed (approved):
Good job.

GerrardE reviewed (approved):
I see you have implemented a complex feature in a few lines of code. This is commendable. I hope you explain the way you implemented this feature to me tomorrow morning. Regards.
